File: backend/src/main/java/ru/rationx/financeapp/services/TransactionService.java
package ru.rationx.financeapp.services;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import ru.rationx.financeapp.models.bank.Bank;
import ru.rationx.financeapp.models.dto.transaction.TransactionDTO;
import ru.rationx.financeapp.models.subject.Subject;
import ru.rationx.financeapp.models.transaction.*;
import ru.rationx.financeapp.repository.TransactionRepository;

import jakarta.persistence.EntityNotFoundException;
import org.springframework.transaction.annotation.Transactional;
import ru.rationx.financeapp.services.exception.NoPermStatusException;
import ru.rationx.financeapp.services.mapper.TransactionMapper;

import java.math.BigDecimal;
import java.security.Principal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.*;

@Service
@RequiredArgsConstructor
@Slf4j
public class TransactionService {

    private final TransactionRepository transactionRepository;
    private final SubjectService subjectService;
    private final BankService bankService;
    private final CategoryService categoryService;
    private final UserService userService;
    private final RegService regService;

    private final TransactionMapper transactionMapper;

    //-----------------------ФИЛЬТРЫ-----------------------------------//

    //Найти все транзакции
    public List<Transaction> getAll() {
        return transactionRepository.findAll();
    }

    /**
     * Получить все транзакции
     */
    public List<Transaction> getAllTransactions() {
        try {
            return transactionRepository.findAll();
        } catch (Exception e) {
            log.error("Error while getting all transactions", e);
            return Collections.emptyList();
        }
    }

    /**
     * Получить транзакцию по ID
     */
    public Transaction getTransactionById(Long id) {
        try {
            Optional<Transaction> transaction = transactionRepository.findById(id);
            return transaction.orElse(null);
        } catch (Exception e) {
            log.error("Error while getting transaction by id: " + id, e);
            return null;
        }
    }

    /**
     * Получить транзакции по статусу
     */
    public List<Transaction> getTransactionsByStatus(TransactionStatus status) {
        try {
            return transactionRepository.findByStatus(status);
        } catch (Exception e) {
            log.error("Error while getting transactions by status: " + status, e);
            return Collections.emptyList();
        }
    }

    /**
     * Получить транзакции по типу
     */
    public List<Transaction> getTransactionsByType(TransactionType type) {
        try {
            return transactionRepository.findByType(type);
        } catch (Exception e) {
            log.error("Error while getting transactions by type: " + type, e);
            return Collections.emptyList();
        }
    }

    /**
     * Получить транзакции по категории
     */
    public List<Transaction> getTransactionsByCategory(Long categoryId) {
        try {
            return transactionRepository.findByCategoryId(categoryId);
        } catch (Exception e) {
            log.error("Error while getting transactions by category id: " + categoryId, e);
            return Collections.emptyList();
        }
    }

    /**
     * Получить транзакции по ИНН получателя
     */
    public List<Transaction> getTransactionsByRecipientInn(String inn) {
        try {
            return transactionRepository.findByRecipientInn(inn);
        } catch (Exception e) {
            log.error("Error while getting transactions by recipient inn: " + inn, e);
            return Collections.emptyList();
        }
    }

    /**
     * Получить транзакции по банку отправителя
     */
    public List<Transaction> getTransactionsBySenderBank(String bank) {
        try {
            return transactionRepository.findBySenderBank(bank);
        } catch (Exception e) {
            log.error("Error while getting transactions by sender bank: " + bank, e);
            return Collections.emptyList();
        }
    }

    /**
     * Получить транзакции по банку получателя
     */
    public List<Transaction> getTransactionsByRecipientBank(String bank) {
        try {
            return transactionRepository.findByRecipientBank(bank);
        } catch (Exception e) {
            log.error("Error while getting transactions by recipient bank: " + bank, e);
            return Collections.emptyList();
        }
    }

    /**
     * Получить транзакции по диапазону дат
     */
    public List<Transaction> getTransactionsByDateRange(LocalDate startDate, LocalDate endDate) {
        try {
            LocalDateTime startDateTime = startDate.atStartOfDay();
            LocalDateTime endDateTime = endDate.atTime(LocalTime.MAX);
            return transactionRepository.findByDateBetween(startDateTime, endDateTime);
        } catch (Exception e) {
            log.error("Error while getting transactions by date range: " + startDate + " - " + endDate, e);
            return Collections.emptyList();
        }
    }

    /**
     * Получить транзакции по диапазону сумм
     */
    public List<Transaction> getTransactionsByAmountRange(BigDecimal minAmount, BigDecimal maxAmount) {
        try {
            return transactionRepository.findByAmountBetween(minAmount, maxAmount);
        } catch (Exception e) {
            log.error("Error while getting transactions by amount range: " + minAmount + " - " + maxAmount, e);
            return Collections.emptyList();
        }
    }

    //-----------------------ДЕЙСТВИЯ С ТРАЗАКЦИЯМИ-----------------------------------//

    //Создать новую транзакцию
    @Transactional
    public Transaction create(TransactionDTO transaction, Principal principal) {

        // Получаем субъекта (отправителя)

        log.info("Запускаем метод перезаписать/создать субъекта (главный). [getOrCreateSubject] ");
        Subject subject = subjectService.getOrCreateSubject(
                transaction.getInn(),
                transaction.getName(),
                transaction.getPersonType(),
                transaction.getAddress(),
                transaction.getPhone()
        );

        log.info("Запускаем метод перезаписать/создать субъекта (recipient). [getOrCreateSubject] ");
        Subject subjectRecipient = subjectService.getOrCreateSubject(
                transaction.getInnRecipient(),
                transaction.getNameRecipient(),
                transaction.getPersonTypeRecipient(),
                transaction.getAddressRecipient(),
                transaction.getRecipientPhoneRecipient()
        );


        log.info("Создаем банк для субъекта ");
        Bank bank = bankService.create(
                transaction.getNameBank(),
                transaction.getRBill(),
                transaction.getBill(),
                subject
        );


        Bank bankRecipient = bankService.create(
                transaction.getNameBankRecip(),
                transaction.getRBillRecip(),
                transaction.getBillRecip(),
                subjectRecipient
        );


        Category category = categoryService.findOrCreateCategory(transaction.getCategory(),
                transaction.getTransactionType());



        Transaction buildTransaction = Transaction.builder()
                .status(TransactionStatus.NEW)
                .dateTime(LocalDateTime.now())
                .subjectSender(subject)
                .subjectGetter(subjectRecipient)
                .senderBank(bank)
                .recipientBank(bankRecipient)
                .comment(transaction.getComment())
                .regTransaction(RegTransaction.builder()
                        .transactionType(transaction.getTypeOperation())
                        .sum(transaction.getSum())
                        .build())
                .category(category)
                .user(userService.getUser(principal.getName()))
                .build();

        transactionRepository.save(buildTransaction);
        log.info("Транзакция успешно собрана без ошибок. ");
        return buildTransaction;

    }

    //Изменение транзакции
    @Transactional
    public Transaction update(Long id, TransactionDTO updatedData) {
        // Получаем транзакцию по id
        Transaction transaction = getById(id);
        if (!transaction.isEditable()) {
            log.error("Возникла ошибка при редактировании, отсутствие прав что бы отредактировать транзакцию. ");
            throw new NoPermStatusException("У вас нет разрешения что бы редактировать транзакцию с такими статусами: " +
                    "{подтвержденная, в обработке, отменена, платеж выполнен, платеж удален, возврат}");
        }

        transactionMapper.updateTransaction(updatedData,transaction);

        // Категория
        Map<String, Object> categoryMap = new HashMap<>();
            categoryMap.put("category",updatedData.getCategory());
            categoryMap.put("type",updatedData.getTypeOperation());

        Category category = categoryService.findOrCreateCategory(updatedData.getCategory(),
                updatedData.getTransactionType());

            categoryService.updateCategoryMap(categoryMap, category);

        // Субъект (главный)
        Map<String,Object> subject0 = new HashMap<>();
            subject0.put("name",updatedData.getName());
            subject0.put("personType",updatedData.getPersonType());
            subject0.put("inn",updatedData.getInn());
            subject0.put("address",updatedData.getAddress());
            subject0.put("phone", updatedData.getPhone());

        Subject sub0 = subjectService.getOrCreateSubject(
                updatedData.getInn(),
                updatedData.getName(),
                updatedData.getPersonType(),
                updatedData.getAddress(),
                updatedData.getPhone()
        );

        subjectService.updateSubject(subject0,sub0);

        // Субъект (дочка)
        Map<String,Object> subject1 = new HashMap<>();
            subject1.put("name",updatedData.getNameRecipient());
            subject1.put("personType",updatedData.getPersonTypeRecipient());
            subject1.put("inn",updatedData.getInnRecipient());
            subject1.put("address",updatedData.getAddressRecipient());
            subject1.put("phone", updatedData.getRecipientPhoneRecipient());

        Subject sub1 = subjectService.getOrCreateSubject(
                updatedData.getInnRecipient(),
                updatedData.getNameRecipient(),
                updatedData.getPersonTypeRecipient(),
                updatedData.getAddressRecipient(),
                updatedData.getRecipientPhoneRecipient()
        );

        subjectService.updateSubject(subject1,sub1);


        // Обновление данных об банках главного субъекта
        Map<String,Object> bank1 = new HashMap<>();
            bank1.put("nameBank",updatedData.getNameBank());
            bank1.put("bill",updatedData.getBill());
            bank1.put("rBill",updatedData.getRBill());
            bank1.put("subj",sub0);
        bankService.updateOrCreateBank(bank1);


        // Обновление данных об банках дочернего субъекта
        Map<String,Object> bank2 = new HashMap<>();
            bank2.put("nameBank",updatedData.getNameBankRecip());
            bank2.put("bill",updatedData.getBillRecip());
            bank2.put("rBill",updatedData.getRBillRecip());
            bank2.put("subj",sub1);
        bankService.updateOrCreateBank(bank2);



        // Обновление данных о регистрах накопления которые приход/расход
        regService.updateReg(transaction,updatedData.getSum(), updatedData.getTransactionType());

        // Ставим статус (опасность налл)
        if (updatedData.getStatus() != null || !updatedData.getStatus().getDescription().isBlank()) {
            log.info("Подтягиваем статус транзакции `{}`...",updatedData.getStatus());
            transaction.setStatus(updatedData.getStatus());
        }

        // Обновляем ссылки
        transaction.setSubjectSender(sub0);
        transaction.setSubjectGetter(sub1);
        transaction.setCategory(category);


        return transactionRepository.save(transaction);
    }

    //Посмотреть по id транзакции
    public Transaction getById(Long id) {
        return transactionRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Транзакция с ID " + id + " не найдена"));
    }

    // найти транзакции по id user
    public List<Transaction> getByUserId(Long id){
        return transactionRepository.findAllByUserId(id);
    }

    //Удаление транзакции
    @Transactional
    public void markAsDeleted(Long id) {
        Transaction transaction = transactionRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Транзакция с ID " + id + " не найдена"));
        
        // Проверка возможности удаления
        if (!transaction.isDeletable()) {
            throw new IllegalStateException("Удаление запрещено для транзакций со статусом " + transaction.getStatus());
        }
        
        // Устанавливаем статус PAYMENT_DELETED вместо физического удаления
        transaction.setStatus(TransactionStatus.PAYMENT_DELETED);
        transactionRepository.save(transaction);
    }

}

File: frontend/src/services/transactionService.ts
import api from './api';

export type PersonType = 'PERSON_TYPE' | 'LEGAL';
export type TransactionType = 'DEBIT' | 'CREDIT';

export interface Transaction {
  id?: number;
  // Отправитель
  personType: PersonType;
  name?: string;
  inn: string;
  address?: string;
  phone?: string;
  
  // Получатель
  personTypeRecipient: PersonType;
  nameRecipient?: string;
  innRecipient: string;
  addressRecipient?: string;
  recipientPhoneRecipient?: string;
  
  // Банк отправителя
  nameBank: string;
  bill?: string;
  rBill?: string;
  
  // Банк получателя
  nameBankRecip: string;
  billRecip: string;
  rBillRecip: string;
  
  // Основные данные транзакции
  comment?: string;
  category: string;
  transactionType: TransactionType;
  sum: number;
  typeOperation: TransactionType;
  transactionDate?: string;
  
  // Маппинг полей для совместимости фронтенда и бэкенда
  amount?: number;
  description?: string;
  categoryId?: number;
  type?: TransactionType;
  status?: string;
  recipientName?: string;
}

const TRANSACTION_ENDPOINTS = {
  BASE: '/api/transactions',
  BY_STATUS: '/api/transactions/status',
  BY_RECIPIENT_INN: '/api/transactions/recipient-inn',
  BY_TYPE: '/api/transactions/type',
  BY_CATEGORY: '/api/transactions/category',
  BY_SENDER_BANK: '/api/transactions/sender-bank',
  BY_RECIPIENT_BANK: '/api/transactions/recipient-bank',
  BY_DATE_RANGE: '/api/transactions/date-range',
  BY_AMOUNT_RANGE: '/api/transactions/amount-range',
};

class TransactionService {
  async getAllTransactions(): Promise<Transaction[]> {
    try {
      const response = await api.get('/api/transactions/get-all-tr');
      console.log('Получены транзакции:', response.data);
      return this._normalizeTransactions(response.data);
    } catch (error) {
      console.error('Failed to fetch transactions:', error);
      throw error;
    }
  }

  async getTransactionsByRecipientInn(inn: string): Promise<Transaction[]> {
    try {
      const response = await api.get(`${TRANSACTION_ENDPOINTS.BY_RECIPIENT_INN}/${inn}`);
      return response.data;
    } catch (error) {
      console.error(`Failed to fetch transactions for recipient INN ${inn}:`, error);
      throw error;
    }
  }

  async getTransactionsByType(type: 'DEBIT' | 'CREDIT'): Promise<Transaction[]> {
    try {
      const response = await api.get(`${TRANSACTION_ENDPOINTS.BY_TYPE}?type=${type}`);
      return response.data;
    } catch (error) {
      console.error(`Failed to fetch ${type} transactions:`, error);
      throw error;
    }
  }

  async getTransactionsByCategory(categoryId: number): Promise<Transaction[]> {
    try {
      const response = await api.get(`${TRANSACTION_ENDPOINTS.BY_CATEGORY}/${categoryId}`);
      return response.data;
    } catch (error) {
      console.error(`Failed to fetch transactions for category ${categoryId}:`, error);
      throw error;
    }
  }

  async getTransactionsBySenderBank(bank: string): Promise<Transaction[]> {
    try {
      const response = await api.get(`${TRANSACTION_ENDPOINTS.BY_SENDER_BANK}/${bank}`);
      return response.data;
    } catch (error) {
      console.error(`Failed to fetch transactions for sender bank ${bank}:`, error);
      throw error;
    }
  }

  async getTransactionsByRecipientBank(bank: string): Promise<Transaction[]> {
    try {
      const response = await api.get(`${TRANSACTION_ENDPOINTS.BY_RECIPIENT_BANK}/${bank}`);
      return response.data;
    } catch (error) {
      console.error(`Failed to fetch transactions for recipient bank ${bank}:`, error);
      throw error;
    }
  }

  async getTransactionsByDateRange(startDate: string, endDate: string): Promise<Transaction[]> {
    try {
      const response = await api.get(
        `${TRANSACTION_ENDPOINTS.BY_DATE_RANGE}?startDate=${startDate}&endDate=${endDate}`
      );
      return response.data;
    } catch (error) {
      console.error(`Failed to fetch transactions between ${startDate} and ${endDate}:`, error);
      throw error;
    }
  }

  async getTransactionsByAmountRange(minAmount: number, maxAmount: number): Promise<Transaction[]> {
    try {
      const response = await api.get(
        `${TRANSACTION_ENDPOINTS.BY_AMOUNT_RANGE}?minAmount=${minAmount}&maxAmount=${maxAmount}`
      );
      return response.data;
    } catch (error) {
      console.error(`Failed to fetch transactions between ${minAmount} and ${maxAmount}:`, error);
      throw error;
    }
  }

  async getTransactionById(id: number): Promise<Transaction> {
    try {
      const response = await api.get(`${TRANSACTION_ENDPOINTS.BASE}/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Failed to fetch transaction with id ${id}:`, error);
      throw error;
    }
  }

  async createTransaction(transaction: Transaction): Promise<Transaction> {
    try {
      // Нормализуем данные перед отправкой на бэкенд
      const transactionForBackend = this._prepareTransactionForBackend(transaction);
      console.log('Отправляем на бэкенд транзакцию:', transactionForBackend);
      const response = await api.post(TRANSACTION_ENDPOINTS.BASE, transactionForBackend);
      return this._normalizeTransactionFromBackend(response.data);
    } catch (error) {
      console.error('Failed to create transaction:', error);
      throw error;
    }
  }

  /**
   * Быстрое создание транзакции с минимальными данными
   */
  async createQuickTransaction(
    amount: number, 
    description: string, 
    categoryId: number, 
    type: TransactionType = 'DEBIT'
  ): Promise<Transaction> {
    const transactionDate = new Date().toISOString().split('T')[0];
    
    // Создаем транзакцию с минимально необходимыми полями
    const transaction: Transaction = {
      // Основные данные
      sum: amount,
      comment: description,
      category: categoryId.toString(),
      transactionType: type,
      typeOperation: type,
      transactionDate: transactionDate,
      
      // Обязательные поля отправителя
      personType: 'PERSON_TYPE',
      inn: '000000000000',
      
      // Обязательные поля получателя
      personTypeRecipient: 'PERSON_TYPE',
      innRecipient: '000000000000',
      
      // Обязательные поля банка отправителя
      nameBank: 'Сбербанк',
      bill: '40817810000000000001',
      rBill: '30101810400000000225',
      
      // Обязательные поля банка получателя
      nameBankRecip: 'Сбербанк',
      billRecip: '40817810000000000002',
      rBillRecip: '30101810400000000226',
      
      // Дополнительные поля для совместимости
      amount,
      description,
      categoryId,
      type
    };
    
    // Отправляем на сервер подготовленные данные
    return this.createTransaction(transaction);
  }

  async updateTransaction(id: number, transaction: Transaction): Promise<Transaction> {
    try {
      const response = await api.put(`${TRANSACTION_ENDPOINTS.BASE}/${id}`, transaction);
      return response.data;
    } catch (error) {
      console.error(`Failed to update transaction with id ${id}:`, error);
      throw error;
    }
  }

  async deleteTransaction(id: number): Promise<boolean> {
    try {
      const response = await api.delete(`${TRANSACTION_ENDPOINTS.BASE}/${id}`);
      return response.status === 200 || response.status === 204;
    } catch (error) {
      console.error(`Failed to delete transaction with id ${id}:`, error);
      throw error;
    }
  }

  // Приватный метод для нормализации транзакций с бэкенда
  private _normalizeTransactions(transactions: any[]): Transaction[] {
    console.log('Нормализация транзакций, исходные данные:', transactions);
    if (!Array.isArray(transactions)) {
      console.error('Ожидался массив транзакций, получено:', transactions);
      return [];
    }
    const normalized = transactions.map(trans => this._normalizeTransactionFromBackend(trans));
    console.log('Нормализованные транзакции:', normalized);
    return normalized;
  }

  // Преобразует транзакцию с бэкенда в формат для фронтенда
  private _normalizeTransactionFromBackend(transaction: any): Transaction {
    if (!transaction) return {} as Transaction;
    
    // Инвертируем типы транзакций:
    // На бэкенде DEBIT - это расход, а CREDIT - это доход
    // Но на фронтенде нужно отображать наоборот
    let transactionType = transaction.transactionType || transaction.type || transaction.typeOperation;
    
    // Инвертируем тип транзакции
    if (transactionType === 'DEBIT') {
      transactionType = 'CREDIT'; // DEBIT (фактический доход) -> отображаем как CREDIT (доход)
    } else if (transactionType === 'CREDIT') {
      transactionType = 'DEBIT'; // CREDIT (фактический расход) -> отображаем как DEBIT (расход)
    }
    
    // Убедимся, что categoryId есть и в правильном формате
    let categoryId = transaction.categoryId;
    if (transaction.category && !categoryId) {
      // Если есть строковая категория, но нет ID - пробуем преобразовать
      const categoryFromString = parseInt(transaction.category);
      if (!isNaN(categoryFromString)) {
        categoryId = categoryFromString;
      }
    }
    
    // Убедимся, что дата в правильном формате
    let transactionDate = transaction.transactionDate;
    if (!transactionDate) {
      // Если нет даты, установим текущую
      transactionDate = new Date().toISOString();
    }
    
    return {
      ...transaction,
      // Маппим поля для совместимости
      amount: transaction.sum || transaction.amount,
      description: transaction.comment || transaction.description,
      type: transactionType,
      // Обеспечиваем совместимость с интерфейсом Transaction
      personType: transaction.personType || 'PERSON_TYPE',
      inn: transaction.inn || '000000000000',
      personTypeRecipient: transaction.personTypeRecipient || 'PERSON_TYPE',
      innRecipient: transaction.innRecipient || '000000000000',
      nameBank: transaction.nameBank || '',
      nameBankRecip: transaction.nameBankRecip || '',
      billRecip: transaction.billRecip || '',
      rBillRecip: transaction.rBillRecip || '',
      typeOperation: transactionType,
      sum: transaction.sum || transaction.amount || 0,
      // Установим и categoryId и category для совместимости
      categoryId: categoryId || (transaction.category ? parseInt(transaction.category) : undefined),
      category: transaction.category || (categoryId ? categoryId.toString() : ''),
      // Установим дату транзакции
      transactionDate: transactionDate,
    };
  }

  // Подготавливает транзакцию для отправки на бэкенд
  private _prepareTransactionForBackend(transaction: Transaction): any {
    const backendTransaction = { ...transaction };
    
    // Устанавливаем все обязательные поля для бэкенда
    backendTransaction.sum = transaction.amount || transaction.sum;
    backendTransaction.comment = transaction.description || transaction.comment;
    
    // Инвертируем тип транзакции для бэкенда
    if (transaction.type === 'DEBIT') {
      backendTransaction.transactionType = 'CREDIT'; // На фронте DEBIT (расход) -> на бэкенде CREDIT
      backendTransaction.typeOperation = 'CREDIT';
    } else if (transaction.type === 'CREDIT') {
      backendTransaction.transactionType = 'DEBIT'; // На фронте CREDIT (доход) -> на бэкенде DEBIT
      backendTransaction.typeOperation = 'DEBIT';
    }
    
    return backendTransaction;
  }
}

export default new TransactionService();
